### PR TITLE
Add MacOS to Github Actions CI tests

### PR DESCRIPTION
**Pull request by kinkie**
Add MacOS/homebrew as a target platform for the GitHub CI tests.
Also improve detection of the number of available cores in buildtests.sh

### CODE REVIEW

**Comment by kinkie** [comment]
@eduard-bagdasaryan , does the changed number of build tests need to be reflected
(if landed) in Anubis? Thanks!

File: .github/workflows/default.yaml [hunk 1]
```diff
@@ -38,7 +38,7 @@ jobs:
           node-version: 16
 
       - name: Checkout Squid sources
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
```
Comment (rousskov) [inline .github/workflows/default.yaml:41]: Please move these checkout version changes into a dedicated PR.

File: .github/workflows/default.yaml [hunk 2]
```diff
@@ -94,7 +94,12 @@ jobs:
     strategy:
       fail-fast: true
       matrix:
-        os: [ ubuntu-22.04 ]
+        os: [ ubuntu-22.04, macos-latest ]
+        layer:
+          - layer-00-default
+          - layer-01-minimal
+          - layer-02-maximus
+          - layer-04-noauth-everything
```
Comment (rousskov) [inline .github/workflows/default.yaml:102]: Please move these out-of-scope changes into a dedicated PR. Since these changes create a lot more CI noise and duplicate information, that dedicated PR should provide data to back its "speed up" claims.

File: .github/workflows/default.yaml [hunk 3]
```diff
@@ -109,10 +114,30 @@ jobs:
...
+            make openldap cyrus-sasl
+
       - name: Checkout sources
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
 
-      - run: ./test-builds.sh
+      - name: Run build on Linux
+        if: runner.os == 'Linux'
+        run: ./test-builds.sh --aggressively-use-config-cache ${{ matrix.layer }}
```
Comment (rousskov) [inline .github/workflows/default.yaml:130]: Please move these out-of-scope changes into a dedicated PR. Since caching had caused a lot of headaches in Jenkins, that dedicated PR should provide data to back the implied "speed up" claims.

**Review by rousskov — CHANGES_REQUESTED** [review]
Thank you very much for making this addition possible!

File: .github/workflows/default.yaml (hunk 2 above)
Comment (kinkie) [reply .github/workflows/default.yaml:102]: https://github.com/kinkie/squid/actions has a lot of output from experiments.

[MacOS serial run with a cold ccache](https://github.com/kinkie/squid/actions/runs/8735961320) - what is currently being done on Linux, takes 1h50
[MacOS serial run without ccache](https://github.com/kinkie/squid/actions/runs/8737350702) takes 1h35. Ccache would help but it is limited to a total 10gb per repository, not enough to hold all the variants we might want to support. With a hot ccache, build time is cut by about 75% ([single layer took 13 minutes](https://github.com/kinkie/squid/actions/runs/8725390026) [no-ccache is 1hr](https://github.com/kinkie/squid/actions/runs/8721931841) ).
[Parallel layer run, no ccache](https://github.com/kinkie/squid/actions/runs/8738417302) takes just short of 20 minutes, an 80% improvement in wall time against the 1h50 baseline

Comment (kinkie) [reply .github/workflows/default.yaml:102]: Note, these times are subject to wide variations, as it depends on which workers the jobs happen to be scheduled and the load of these workers, but I hope I managed to convince you that they're beneficial.

CPU time increases are minimal, mostly related to the duplicate time spent installing dependencies

Comment (kinkie) [reply .github/workflows/default.yaml:102]: Extracted in PR #1787

File: .github/workflows/default.yaml (hunk 1 above)
Comment (kinkie) [reply .github/workflows/default.yaml:41]: Done in PR #1786

File: .github/workflows/default.yaml (hunk 3 above)
Comment (kinkie) [reply .github/workflows/default.yaml:130]: This option refers to config.cache (used by configure); using it could be harmful only if the system were to change between runs, but since runners get wiped and rebuilt on each run, no harm is possible.
If anything, if the per-layer parallelisation is accepted, this is useless because there is no chance to
reuse the cache

Comment (rousskov) [reply .github/workflows/default.yaml:130]: While your response raises more red flags, I will pass this opportunity to discuss these out-of-scope changes in this PR. If you are sure these changes are a good idea, please see my change request for the next steps.

**Comment by eduard-bagdasaryan** [comment]
> does the changed number of build tests need to be reflected (if landed) in Anubis? 

The 'staging_checks' Anubis' parameter should be adjusted if the number of required checks changes. So, after marking these checks as 'required' (in the protected branch GitHub configuration), we should fix the parameter respectively. It is 17 now: 4 GitHub checks + 12 Jenkins checks + approval.

**Comment by rousskov** [comment]
> So, after marking these checks as 'required' (in the protected branch GitHub configuration), we should fix the parameter respectively

Just to add another wrinkle: To merge a PR that changes the build-test matrix, one also has to _remove_ the "required" mark from the existing build-test matrix entries that are going away _and_ adjust Anubis configuration to match.

Right now, there is only one such entry; you can see it with a yellow dot in this and any similar PR (because GitHub does not start this expected/required test after PR changes):

```
* build-tests (ubuntu-22.04) Expected — Waiting for status to be reported 
```

So, if we do not want to brute-force anything or risk merging PRs that break tests, then the procedure may look something like this (this is an untested sketch):

1. Get a tests-changing PR approved. Check that all new/remaining tests are green and all going-away tests are yellow. 
2. Mark the new test(s) that are being added as required in GitHub protected branch configuration(s).
3. _After_ item 2, mark the existing test(s) that are going away as optional in GitHub protected branch configuration(s).
4. [Adjust](https://github.com/squid-cache/squid/pull/1785#issuecomment-2066472317) Anubis configuration to match GitHub configuration.
5. Add a GitHub comment summarizing the above changes and wait for Anubis to merge the PR.
6. Merge master into other pending PRs as needed. They will not be auto-merged until their branches get new tests (because GitHub testing requirements have changed).

File: .github/workflows/default.yaml [hunk 4]
```diff
@@ -109,10 +114,30 @@ jobs:
...
+
+      - name: Run build on MacOS
+        if: runner.os == 'macOS'
+        run: |
+          eval `brew shellenv`
+          export PKG_CONFIG_PATH="$HOMEBREW_PREFIX/opt/openldap/lib/pkgconfig:$HOMEBREW_PREFIX/opt/cyrus-sasl/lib/pkgconfig"
+          export GETOPT="$HOMEBREW_PREFIX/opt/gnu-getopt/bin/getopt"
+          export pjobs="-j`nproc`"
+          export MAKE="$HOMEBREW_PREFIX/bin/gmake ${pjobs}"
```
Comment (rousskov) [inline .github/workflows/default.yaml:139]: I assume that `-j` is an optimization rather than a requirement to build on MacOS. Let's not increase the difference between Linux and MacOS builds in this PR where we do not have to:

```suggestion
          export MAKE="$HOMEBREW_PREFIX/bin/gmake"
```

If my assumption is wrong, please add a (YAML) comment to document what is going on in this special case.

File: .github/workflows/default.yaml [hunk 5]
```diff
@@ -94,7 +94,12 @@ jobs:
     strategy:
       fail-fast: true
       matrix:
-        os: [ ubuntu-22.04 ]
+        os: [ ubuntu-22.04, macos-latest ]
```
Comment (rousskov) [inline .github/workflows/default.yaml:97]: Currently, `macos-14` [is the same](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories) as `macos-latest`, but we should use the former for the same reasons we are currently using `ubuntu-22.04` rather than `ubuntu-latest`.

```suggestion
        os: [ ubuntu-22.04, macos-14 ]
```

**Review by rousskov — CHANGES_REQUESTED** [review]

File: .github/workflows/default.yaml (hunk 2 above)
Comment (kinkie) [reply .github/workflows/default.yaml]: Removed from here

**Comment by kinkie** [comment]
As @eduard-bagdasaryan mentioned, to land this we will need to change Anubis' configuration

File: .github/workflows/default.yaml [hunk 6]
```diff
@@ -109,10 +109,29 @@ jobs:
...
+        run: ./test-builds.sh --aggressively-use-config-cache
+
+      - name: Run build on MacOS
+        if: runner.os == 'macOS'
+        run: |
+          eval `brew shellenv`
+          export PKG_CONFIG_PATH="$HOMEBREW_PREFIX/opt/openldap/lib/pkgconfig:$HOMEBREW_PREFIX/opt/cyrus-sasl/lib/pkgconfig"
+          export GETOPT="$HOMEBREW_PREFIX/opt/gnu-getopt/bin/getopt"
+          export MAKE="$HOMEBREW_PREFIX/bin/gmake"
+          ./test-builds.sh --aggressively-use-config-cache
```
Comment (rousskov) [inline .github/workflows/default.yaml:134]: Just in case it was not clear before, the change request about caching applies to this caller as well.

**Review by rousskov — CHANGES_REQUESTED** [review]
Please do not forget to fix this PR description. I cannot do that easily right now.

File: .github/workflows/default.yaml [hunk 7]
```diff
@@ -109,10 +109,29 @@ jobs:
...
+            gnu-getopt gnu-sed grep \
+            make openldap cyrus-sasl
+
       - name: Checkout sources
         uses: actions/checkout@v4
 
-      - run: ./test-builds.sh
+      - name: Run build on Linux
+        if: runner.os == 'Linux'
+        run: ./test-builds.sh --aggressively-use-config-cache
```
[the remark this answers is not in the document]
Comment (kinkie) [reply .github/workflows/default.yaml:125]: removed

File: .github/workflows/default.yaml [hunk 8]
```diff
@@ -109,10 +109,35 @@ jobs:
...
+        run: ./test-builds.sh
+
+      - name: Run build on MacOS
+        if: runner.os == 'macOS'
+        run: |
+          eval `brew shellenv`
+          export PKG_CONFIG_PATH="$HOMEBREW_PREFIX/lib/pkgconfig:$HOMEBREW_PREFIX/opt/openldap/lib/pkgconfig:$HOMEBREW_PREFIX/opt/cyrus-sasl/lib/pkgconfig"
+          export GETOPT="$HOMEBREW_PREFIX/opt/gnu-getopt/bin/getopt"
+          if gcc --version | grep -q clang ; then
+            export CFLAGS="-Wno-compound-token-split-by-macro${CFLAGS+ $CFLAGS}" # needed fir ltdl
```
Comment (rousskov) [inline .github/workflows/default.yaml:134]:
```
${name+word}
${name:+word}
       If name is set, or in the second form is non-null, then
       substitute word; otherwise substitute nothing.
```

Don't we want the "second form" here (to avoid implying that we want to add a null/empty CFLAGS value)?

```suggestion
            export CFLAGS="-Wno-compound-token-split-by-macro${CFLAGS:+ $CFLAGS}" # needed fir ltdl
```

Added bonus: More folks might be familiar with `${x:+z}` notation than with `${x+z}`. FWIW, it took me a while to find documentation for the latter.

Same for the other two occurrences below, of course.

File: .github/workflows/default.yaml [hunk 9]
```diff
@@ -109,10 +109,35 @@ jobs:
...
+          fi
+          export pjobs="-j`nproc`"
+          export MAKE="$HOMEBREW_PREFIX/bin/gmake"
+          export CPPFLAGS="-I$HOMEBREW_PREFIX/include${CPPFLAGS+ $CPPFLAGS}" # openssl fails without this
```
Comment (rousskov) [inline .github/workflows/default.yaml:138]: Is `-I$HOMEBREW_PREFIX/include` missing from the OpenSSL pkg-config configuration provided by homebrew or is it there, but we are mishandling it somehow? Just curious...

File: .github/workflows/default.yaml [hunk 10]
```diff
@@ -109,10 +109,35 @@ jobs:
...
+      - name: Run build on MacOS
+        if: runner.os == 'macOS'
+        run: |
+          eval `brew shellenv`
+          export PKG_CONFIG_PATH="$HOMEBREW_PREFIX/lib/pkgconfig:$HOMEBREW_PREFIX/opt/openldap/lib/pkgconfig:$HOMEBREW_PREFIX/opt/cyrus-sasl/lib/pkgconfig"
+          export GETOPT="$HOMEBREW_PREFIX/opt/gnu-getopt/bin/getopt"
+          if gcc --version | grep -q clang ; then
+            export CFLAGS="-Wno-compound-token-split-by-macro${CFLAGS+ $CFLAGS}" # needed fir ltdl
+          fi
+          export pjobs="-j`nproc`"
```
Comment (rousskov) [inline .github/workflows/default.yaml:136]: Please document or remove as unused

```suggestion
```

File: .github/workflows/default.yaml [hunk 11]
```diff
@@ -109,10 +109,35 @@ jobs:
           sudo apt-get --quiet=2 build-dep squid
           sudo apt-get --quiet=2 install linuxdoc-tools libtool-bin
 
+      - name: Install prerequisite MacOS packages
+        if: runner.os == 'macOS'
+        run: |
+          brew install \
+            automake coreutils cppunit gawk \
+            gnu-getopt gnu-sed grep libtool \
+            make openldap openssl@3 cyrus-sasl
```
Comment (rousskov) [inline .github/workflows/default.yaml:118]: Do we have to specify OpenSSL version?

**Review by rousskov — CHANGES_REQUESTED** [review]
> I believe it's ready for merge, only waiting for review

Not ready -- PR description is still stale.

File: .github/workflows/default.yaml (hunk 9 above)
Comment (rousskov) [reply .github/workflows/default.yaml:138]: Ouch. I am surprised such a bug has not been fixed upstream yet.

If OpenSSL v1 pkg-config configuration has similar problems, then I suggest:

```suggestion
           # This is done to keep in sync with LDFLAGS below
          export CPPFLAGS="-I$HOMEBREW_PREFIX/include${CPPFLAGS+ $CPPFLAGS}"
           # `pkg-config --libs openssl` points to a wrong directory (missing lib/)
```

Reordering CPPFLAGS and LDFLAGS blobs (and s/below/above/) is probably a good idea in that case.

Otherwise, if we can stick with OpenSSL v1 (if it is available by default and "works" without complications), then we should probably test with that version for now.

File: .github/workflows/default.yaml [hunk 12]
```diff
@@ -109,10 +109,35 @@ jobs:
...
-      - run: ./test-builds.sh
+      - name: Run build on Linux
+        if: runner.os == 'Linux'
+        run: ./test-builds.sh
+
+      - name: Run build on MacOS
+        if: runner.os == 'macOS'
+        run: |
+          eval `brew shellenv`
+          export PKG_CONFIG_PATH="$HOMEBREW_PREFIX/lib/pkgconfig:$HOMEBREW_PREFIX/opt/openldap/lib/pkgconfig:$HOMEBREW_PREFIX/opt/cyrus-sasl/lib/pkgconfig"
```
Comment (rousskov) [inline .github/workflows/default.yaml:131]: I would split this into package-specific lines. The code becomes much more readable (and easier to maintain) that way IMO. For example:

```suggestion
          export PKG_CONFIG_PATH="$HOMEBREW_PREFIX/lib/pkgconfig"
          PKG_CONFIG_PATH="$PKG_CONFIG_PATH:$HOMEBREW_PREFIX/opt/openldap/lib/pkgconfig"
          PKG_CONFIG_PATH="$PKG_CONFIG_PATH:$HOMEBREW_PREFIX/opt/cyrus-sasl/lib/pkgconfig"
```

File: .github/workflows/default.yaml (hunk 11 above)
Comment (kinkie) [reply .github/workflows/default.yaml]: removing version specifier

File: .github/workflows/default.yaml (hunk 12 above)
Comment (kinkie) [reply .github/workflows/default.yaml]: I don't think it helps much but it doesn't hurt either, so adopting a
even more readable version to advance the PR

File: .github/workflows/default.yaml (hunk 8 above)
Comment (kinkie) [reply .github/workflows/default.yaml]: It is more future proof, in that it will handle CFLAGS lines being added earlier in the script cleanly, so there is some value.
Adopting the :+ syntax

File: .github/workflows/default.yaml (hunk 10 above)
Comment (kinkie) [reply .github/workflows/default.yaml]: done

**Comment by kinkie** [comment]
Replying to https://github.com/squid-cache/squid/pull/1785#discussion_r1578683277:

Reworded in a slightly different way.

File: test-suite/buildtest.sh [hunk 13]
```diff
@@ -22,19 +22,25 @@ if [ -n "$cache_file" ]; then
     configcache="--cache-file=$cache_file"
 fi
 
-#if we are on Linux, let's try parallelizing
+# let's try parallelizing
+if [ -z "$pjobs" ] && which nproc >/dev/null 2>&1 ; then
```
[the remark this answers is not in the document]
Comment (kinkie) [reply test-suite/buildtest.sh:26]: The problem is not that running nproc cheap, but that it might not be installed.
nproc is part of GNU coreutils, and it's not a given it will be installed on a non-Linux system.
Your proposed approach would result in a shell-emitted error message about nproc not
being available - everything would work just fine, but the error message might be confusing

File: test-suite/buildtest.sh [hunk 14]
```diff
@@ -22,19 +22,25 @@ if [ -n "$cache_file" ]; then
...
+    ncpus=`nproc`
+    pjobs="-j${ncpus}"
+fi
+# no nproc, Linux-specific detection
```
[the remark this answers is not in the document]
Comment (kinkie) [reply test-suite/buildtest.sh:30]: /proc/cpuinfo IS Linux specific, no other OS has it.
I'm removing the comment since you don't like it, but ¯\_(ツ)_/¯

File: test-suite/buildtest.sh (hunk 13 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply test-suite/buildtest.sh:26]: The error message will not be visible to anybody AFAICT due to output redirection. Am I missing something?

[the remark this answers is not in the document]
Comment (kinkie) [reply test-suite/buildtest.sh:26]: I had tested it. The error message was visible, I think because it is emitted by the shell itself, before output redirection was done

[the remark this answers is not in the document]
Comment (rousskov) [reply test-suite/buildtest.sh:26]: This redirection is a standard shell trick we are using elsewhere (including PR code), but I tested it just to make sure. The error message is not visible in all cases I have tested, on two OSes (macOS and Linux):

```
% nproc
zsh: command not found: nproc

% if [ -z "$pjobs" ] && nproc > /dev/null 2>&1 ; then echo TRYING; else echo NOT TRYING; fi
NOT TRYING

% prjobs=12
% if [ -z "$pjobs" ] && nproc > /dev/null 2>&1 ; then echo TRYING; else echo NOT TRYING; fi
NOT TRYING
```

```
$ nproc
16

$ if [ -z "$pjobs" ] && nproc > /dev/null 2>&1 ; then echo TRYING; else echo NOT TRYING; fi
TRYING
$ if [ -z "$pjobs" ] && nprocXXX > /dev/null 2>&1 ; then echo TRYING; else echo NOT TRYING; fi
NOT TRYING

$ pjobs=12
$ if [ -z "$pjobs" ] && nprocXXX > /dev/null 2>&1 ; then echo TRYING; else echo NOT TRYING; fi
NOT TRYING
$ if [ -z "$pjobs" ] && nproc > /dev/null 2>&1 ; then echo TRYING; else echo NOT TRYING; fi
NOT TRYING
```

----

If you need more reasons to avoid `which` besides the original change request, some folks [assert](https://stackoverflow.com/a/677212) that "Many operating systems have a `which` that doesn't even set an exit status".

**Comment by kinkie** [comment]
> If OpenSSL v1 pkg-config configuration has similar problems, then I suggest:

No, openssl@1 does not have the same problem

File: .github/workflows/default.yaml [hunk 15]
```diff
@@ -109,10 +109,40 @@ jobs:
...
+      - name: Run build on MacOS
+        if: runner.os == 'macOS'
+        run: |
+          eval `brew shellenv`
+          PKG_CONFIG_PATH="$HOMEBREW_PREFIX/lib/pkgconfig"
+          PKG_CONFIG_PATH="$PKG_CONFIG_PATH:$HOMEBREW_PREFIX/opt/openldap/lib/pkgconfig"
+          PKG_CONFIG_PATH="$PKG_CONFIG_PATH:$HOMEBREW_PREFIX/opt/cyrus-sasl/lib/pkgconfig"
+          export PKG_CONFIG_PATH
+          export GETOPT="$HOMEBREW_PREFIX/opt/gnu-getopt/bin/getopt"
+          if gcc --version | grep -q clang ; then
```
Comment (rousskov) [inline .github/workflows/default.yaml:136]: Is it possible to test `c++` here instead of `gcc`? It would make this hack less odd/controversial and more readable:

```suggestion
          if c++ --version | grep -q clang ; then
```

P.S. I am surprised this is how we are supposed to detect `-Wno-compound-token-split-by-macro` option support, but I am not going to fight for better detection in this rather constraint environment, despite the fact that, in a constraint environment like this, we do not need to be testing which compiler we were given -- we should know/assume the right answer!

**Review by rousskov — COMMENTED** [review]
No changes in my `-0` position since the last review. Thank you for addressing my earlier concerns.

**Review by rousskov — APPROVED** [review]

File: .github/workflows/default.yaml (hunk 15 above)
Comment (kinkie) [reply .github/workflows/default.yaml:136]: We can test c++, but it's really a C problem.
The environment we are targeting is:
```
muno master/src ‹master› » gcc --version
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin23.4.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

Using c++ has no real advantage:
```
c++ --version
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin23.4.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

If anything, we might want to try "cc" out instead of "gcc", but then Xcode installs a gcc-compatible binary.


> we do not need to be testing which compiler we were given -- we should know/assume the right answer
Sure, we can unconditionally define the flag. The point is: without it, libltdl fails the build

Comment (rousskov) [reply .github/workflows/default.yaml]: Using correct vendor-neutral/non-gcc compiler name would address this change request. If that name should be cc rather than c++, please use cc, of course.

Comment (rousskov) [reply .github/workflows/default.yaml]: And using correct unconditional flags is also fine.

**Review by rousskov — APPROVED** [review]
